File: swift/Sources/Rezel/Grammars/PythonGrammar.swift
import Foundation

// MARK: - Character constants

private let newline = 10
private let carriageReturn = 13
private let space = 32
private let tab = 9
private let hash = 35
private let parenOpen = 40
private let dot = 46
private let braceOpen = 123
private let braceClose = 125
private let singleQuote = 39
private let doubleQuote = 34
private let backslash = 92
private let letter_o = 111
private let letter_x = 120
private let letter_N = 78
private let letter_u = 117
private let letter_U = 85

// MARK: - Flags

private let cx_Bracketed = 1
private let cx_String = 2
private let cx_DoubleQuote = 4
private let cx_Long = 8
private let cx_Raw = 16
private let cx_Format = 32

// MARK: - Helpers

private func isLineBreak(_ ch: Int) -> Bool {
	ch == newline || ch == carriageReturn
}

private func isHex(_ ch: Int) -> Bool {
	(ch >= 48 && ch <= 57) || (ch >= 65 && ch <= 70) || (ch >= 97 && ch <= 102)
}

private func isWordChar(_ ch: Int) -> Bool {
	guard let scalar = UnicodeScalar(ch) else { return false }
	return scalar.properties.isAlphabetic || (ch >= 48 && ch <= 57) || ch == 95
}

private func countIndent(_ s: String) -> Int {
	var depth = 0
	for ch in s.unicodeScalars {
		if ch == "\t" {
			depth += 8 - (depth % 8)
		} else {
			depth += 1
		}
	}
	return depth
}

private func skipEscape(_ input: InputStream, _ ch: Int) {
	if ch == letter_o {
		var i = 0
		while i < 2, input.next >= 48, input.next <= 55 {
			input.advance(); i += 1
		}
	} else if ch == letter_x {
		var i = 0
		while i < 2, isHex(input.next) {
			input.advance(); i += 1
		}
	} else if ch == letter_u {
		var i = 0
		while i < 4, isHex(input.next) {
			input.advance(); i += 1
		}
	} else if ch == letter_U {
		var i = 0
		while i < 8, isHex(input.next) {
			input.advance(); i += 1
		}
	} else if ch == letter_N {
		if input.next == braceOpen {
			input.advance()
			while input.next >= 0, input.next != braceClose, input.next != singleQuote,
			      input.next != doubleQuote, input.next != newline
			{
				input.advance()
			}
			if input.next == braceClose { input.advance() }
		}
	}
}

// MARK: - Context

private class PythonContext {
	let parent: PythonContext?
	let indent: Int
	let flags: Int
	let hashValue: Int

	init(parent: PythonContext?, indent: Int, flags: Int) {
		self.parent = parent
		self.indent = indent
		self.flags = flags
		let parentHash = parent?.hashValue ?? 0
		hashValue = parentHash &+ (parentHash << 8) &+ indent &+ (indent << 4) &+ flags &+ (flags << 6)
	}
}

private nonisolated(unsafe) let pythonTopContext = PythonContext(parent: nil, indent: 0, flags: 0)

// MARK: - Term name sets / maps

private let pythonBracketedNames: Set<String> = [
	"ParenthesizedExpression", "TupleExpression", "ComprehensionExpression",
	"importList", "ArgList", "ParamList",
	"ArrayExpression", "ArrayComprehensionExpression", "subscript",
	"SetExpression", "SetComprehensionExpression",
	"FormatString", "FormatReplacement", "nestedFormatReplacement",
	"DictionaryExpression", "DictionaryComprehensionExpression",
	"SequencePattern", "MappingPattern", "PatternArgList", "TypeParamList",
]

// The opening bracket tokens `(`, `[`, `{` are anonymous terms named by their literal value.
private let pythonBracketOpenNames: Set<String> = ["\"(\"", "\"[\"", "\"{\""]

private let pythonStringFlags: [String: Int] = [
	"stringStart": cx_String,
	"stringStartD": cx_String | cx_DoubleQuote,
	"stringStartL": cx_String | cx_Long,
	"stringStartLD": cx_String | cx_Long | cx_DoubleQuote,
	"stringStartR": cx_String | cx_Raw,
	"stringStartRD": cx_String | cx_Raw | cx_DoubleQuote,
	"stringStartRL": cx_String | cx_Raw | cx_Long,
	"stringStartRLD": cx_String | cx_Raw | cx_Long | cx_DoubleQuote,
	"stringStartF": cx_String | cx_Format,
	"stringStartFD": cx_String | cx_Format | cx_DoubleQuote,
	"stringStartFL": cx_String | cx_Format | cx_Long,
	"stringStartFLD": cx_String | cx_Format | cx_Long | cx_DoubleQuote,
	"stringStartFR": cx_String | cx_Format | cx_Raw,
	"stringStartFRD": cx_String | cx_Format | cx_Raw | cx_DoubleQuote,
	"stringStartFRL": cx_String | cx_Format | cx_Raw | cx_Long,
	"stringStartFRLD": cx_String | cx_Format | cx_Raw | cx_Long | cx_DoubleQuote,
]

// MARK: - Context tracker

nonisolated(unsafe) let pythonTrackIndent = ContextTracker(
	start: pythonTopContext as Any,
	shift: { context, term, stack, input in
		guard let cx = context as? PythonContext else { return context }
		let name = stack.parser.termNames?[term] ?? ""
		if name == "indent" {
			let indentStr = input.read(from: input.pos, to: stack.pos)
			return PythonContext(parent: cx, indent: countIndent(indentStr), flags: 0) as Any
		}
		if name == "dedent" {
			return (cx.parent ?? pythonTopContext) as Any
		}
		if pythonBracketOpenNames.contains(name) || name == "replacementStart" {
			return PythonContext(parent: cx, indent: 0, flags: cx_Bracketed) as Any
		}
		if let flags = pythonStringFlags[name] {
			return PythonContext(parent: cx, indent: 0, flags: flags | (cx.flags & cx_Bracketed)) as Any
		}
		return context
	},
	reduce: { context, term, stack, _ in
		guard let cx = context as? PythonContext else { return context }
		let name = stack.parser.termNames?[term] ?? ""
		if (cx.flags & cx_Bracketed) != 0, pythonBracketedNames.contains(name) {
			return (cx.parent ?? pythonTopContext) as Any
		}
		if name == "String" || name == "FormatString", (cx.flags & cx_String) != 0 {
			return (cx.parent ?? pythonTopContext) as Any
		}
		return context
	},
	hash: { context in
		(context as? PythonContext)?.hashValue ?? 0
	}
)

// MARK: - External tokenizers

private func makePythonExternalTokenizer(name: String, terms: [String: Int]) -> TokenizerProtocol {
	if name == "newlines" {
		let eofT = terms["eof"]!
		let newlineBracketedT = terms["newlineBracketed"]!
		let blankLineStartT = terms["blankLineStart"]!
		let newlineT = terms["newline"]!
		return ExternalTokenizer({ input, stack in
			let flags = (stack.context as? PythonContext)?.flags ?? 0
			if input.next < 0 {
				input.acceptToken(eofT)
			} else if (flags & cx_Bracketed) != 0 {
				if isLineBreak(input.next) {
					input.acceptToken(newlineBracketedT, endOffset: 1)
				}
			} else {
				let prev = input.peek(-1)
				if prev < 0 || isLineBreak(prev), stack.canShift(blankLineStartT) {
					var spaces = 0
					while input.next == space || input.next == tab {
						input.advance()
						spaces += 1
					}
					if input.next == newline || input.next == carriageReturn || input.next == hash {
						input.acceptToken(blankLineStartT, endOffset: -spaces)
					}
				} else if isLineBreak(input.next) {
					input.acceptToken(newlineT, endOffset: 1)
				}
			}
		}, contextual: true)
	}

	if name == "indentation" {
		let dedentT = terms["dedent"]!
		let indentT = terms["indent"]!
		return ExternalTokenizer { input, stack in
			guard let cx = stack.context as? PythonContext else { return }
			if cx.flags != 0 { return }
			let prev = input.peek(-1)
			if prev == newline || prev == carriageReturn {
				var depth = 0
				var chars = 0
				while true {
					if input.next == space {
						depth += 1
					} else if input.next == tab {
						depth += 8 - (depth % 8)
					} else {
						break
					}
					input.advance()
					chars += 1
				}
				if depth != cx.indent,
				   input.next != newline, input.next != carriageReturn, input.next != hash
				{
					if depth < cx.indent {
						input.acceptToken(dedentT, endOffset: -chars)
					} else {
						input.acceptToken(indentT)
					}
				}
			}
		}
	}

	if name == "legacyPrint" {
		let printKeywordT = terms["printKeyword"]!
		let printChars: [Int] = [112, 114, 105, 110, 116] // "print"
		return ExternalTokenizer { input, _ in
			for i in 0 ..< 5 {
				if input.next != printChars[i] { return }
				input.advance()
			}
			if isWordChar(input.next) { return }
			var off = 0
			while true {
				let next = input.peek(off)
				if next == space || next == tab { off += 1; continue }
				if next != parenOpen, next != dot, next != newline,
				   next != carriageReturn, next != hash
				{
					input.acceptToken(printKeywordT)
				}
				return
			}
		}
	}

	if name == "strings" {
		let stringEndT = terms["stringEnd"]!
		let stringContentT = terms["stringContent"]!
		let escapeT = terms["Escape"]!
		let replacementStartT = terms["replacementStart"]!
		return ExternalTokenizer { input, stack in
			guard let cx = stack.context as? PythonContext else { return }
			let flags = cx.flags
			let quote = (flags & cx_DoubleQuote) != 0 ? doubleQuote : singleQuote
			let isLong = (flags & cx_Long) != 0
			let escapes = (flags & cx_Raw) == 0
			let format = (flags & cx_Format) != 0

			let start = input.pos
			scan: while true {
				if input.next < 0 {
					break scan
				} else if format, input.next == braceOpen {
					if input.peek(1) == braceOpen {
						input.advance(2)
					} else {
						if input.pos == start {
							input.acceptToken(replacementStartT, endOffset: 1)
							return
						}
						break scan
					}
				} else if escapes, input.next == backslash {
					if input.pos == start {
						input.advance()
						let escaped = input.next
						if escaped >= 0 {
							input.advance()
							skipEscape(input, escaped)
						}
						input.acceptToken(escapeT)
						return
					}
					break scan
				} else if input.next == backslash, !escapes, input.peek(1) > -1 {
					input.advance(2)
				} else if input.next == quote, !isLong || (input.peek(1) == quote && input.peek(2) == quote) {
					if input.pos == start {
						input.acceptToken(stringEndT, endOffset: isLong ? 3 : 1)
						return
					}
					break scan
				} else if input.next == newline {
					if isLong {
						input.advance()
					} else if input.pos == start {
						input.acceptToken(stringEndT)
						return
					}
					break scan
				} else {
					input.advance()
				}
			}
			if input.pos > start { input.acceptToken(stringContentT) }
		}
	}

	fatalError("Unknown Python external tokenizer: \(name)")
}

// MARK: - Highlighting

private nonisolated(unsafe) let definition = hlTags["definition"] as! (Tag) -> Tag
private nonisolated(unsafe) let functionMod = hlTags["function"] as! (Tag) -> Tag
private nonisolated(unsafe) let specialMod = hlTags["special"] as! (Tag) -> Tag

nonisolated(unsafe) let pythonHighlighting = styleTags([
	"async \"*\" \"**\" FormatConversion FormatSpec": hlTags["modifier"] as Any,
	"for while if elif else try except finally return raise break continue with pass assert await yield match case": hlTags["controlKeyword"] as Any,
	"in not and or is del": hlTags["operatorKeyword"] as Any,
	"from def class global nonlocal lambda": hlTags["definitionKeyword"] as Any,
	"import": hlTags["moduleKeyword"] as Any,
	"with as print": hlKeyword,
	"Boolean": hlTags["bool"] as Any,
	"None": hlTags["null"] as Any,
	"VariableName": hlTags["variableName"] as Any,
	"CallExpression/VariableName": functionMod(hlTags["variableName"] as! Tag),
	"FunctionDefinition/VariableName": functionMod(definition(hlTags["variableName"] as! Tag)),
	"ClassDefinition/VariableName": definition(hlTags["className"] as! Tag),
	"PropertyName": hlPropertyName,
	"CallExpression/MemberExpression/PropertyName": functionMod(hlPropertyName),
	"Comment": hlTags["lineComment"] as Any,
	"Number": hlNumber,
	"String": hlString,
	"FormatString": specialMod(hlString),
	"Escape": hlTags["escape"] as Any,
	"UpdateOp": hlTags["updateOperator"] as Any,
	"ArithOp!": hlTags["arithmeticOperator"] as Any,
	"BitOp": hlTags["bitwiseOperator"] as Any,
	"CompareOp": hlTags["compareOperator"] as Any,
	"AssignOp": hlTags["definitionOperator"] as Any,
	"Ellipsis": hlPunctuation,
	"At": hlMeta,
	"( )": hlTags["paren"] as Any,
	"[ ]": hlTags["squareBracket"] as Any,
	"{ }": hlTags["brace"] as Any,
	".": hlTags["derefOperator"] as Any,
	", ;": hlTags["separator"] as Any,
])

// MARK: - Parser

public nonisolated(unsafe) let pythonParser: LRParser = {
	let externals: [String: TokenizerProtocol] = [
		"newlines": makePythonExternalTokenizer(name: "newlines", terms: PythonParserData.termTable),
		"indentation": makePythonExternalTokenizer(name: "indentation", terms: PythonParserData.termTable),
		"legacyPrint": makePythonExternalTokenizer(name: "legacyPrint", terms: PythonParserData.termTable),
		"strings": makePythonExternalTokenizer(name: "strings", terms: PythonParserData.termTable),
	]
	let spec = PythonParserData.makeSpec(
		externals: externals,
		propSources: [pythonHighlighting],
		context: pythonTrackIndent
	)
	return LRParser(spec: spec)
}()

File: dotnet/Rezel/Grammars/PythonGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class PythonGrammar
{
    private const int Newline = 10;
    private const int CarriageReturn = 13;
    private const int Space = 32;
    private const int Tab = 9;
    private const int Hash = 35;
    private const int ParenOpen = 40;
    private const int Dot = 46;
    private const int BraceOpen = 123;
    private const int BraceClose = 125;
    private const int SingleQuote = 39;
    private const int DoubleQuote = 34;
    private const int Backslash = 92;
    private const int LetterO = 111;
    private const int LetterX = 120;
    private const int LetterN = 78;
    private const int LetterU = 117;
    private const int LetterBigU = 85;

    private const int CxBracketed = 1;
    private const int CxString = 2;
    private const int CxDoubleQuote = 4;
    private const int CxLong = 8;
    private const int CxRaw = 16;
    private const int CxFormat = 32;

    private class PyContext
    {
        public PyContext? Parent;
        public int Indent;
        public int Flags;
        public int Hash;

        public PyContext(PyContext? parent, int indent, int flags)
        {
            Parent = parent;
            Indent = indent;
            Flags = flags;
            Hash = (parent != null ? parent.Hash + (parent.Hash << 8) : 0) + indent + (indent << 4) + flags + (flags << 6);
        }
    }

    private static readonly PyContext TopIndent = new(null, 0, 0);

    private static bool IsLineBreak(int ch) => ch == Newline || ch == CarriageReturn;
    private static bool IsHex(int ch) => (ch >= 48 && ch <= 57) || (ch >= 65 && ch <= 70) || (ch >= 97 && ch <= 102);

    private static int CountIndent(string spaceStr)
    {
        var depth = 0;
        for (var i = 0; i < spaceStr.Length; i++)
            depth += spaceStr[i] == Tab ? 8 - (depth % 8) : 1;
        return depth;
    }

    private static Dictionary<string, int> _stringFlags = new();

    private static void BuildStringFlags(Dictionary<string, int> terms)
    {
        if (_stringFlags.Count > 0) return;
        _stringFlags["stringStart"] = 0 | CxString;
        _stringFlags["stringStartD"] = CxDoubleQuote | CxString;
        _stringFlags["stringStartL"] = CxLong | CxString;
        _stringFlags["stringStartLD"] = CxLong | CxDoubleQuote | CxString;
        _stringFlags["stringStartR"] = CxRaw | CxString;
        _stringFlags["stringStartRD"] = CxRaw | CxDoubleQuote | CxString;
        _stringFlags["stringStartRL"] = CxRaw | CxLong | CxString;
        _stringFlags["stringStartRLD"] = CxRaw | CxLong | CxDoubleQuote | CxString;
        _stringFlags["stringStartF"] = CxFormat | CxString;
        _stringFlags["stringStartFD"] = CxFormat | CxDoubleQuote | CxString;
        _stringFlags["stringStartFL"] = CxFormat | CxLong | CxString;
        _stringFlags["stringStartFLD"] = CxFormat | CxLong | CxDoubleQuote | CxString;
        _stringFlags["stringStartFR"] = CxFormat | CxRaw | CxString;
        _stringFlags["stringStartFRD"] = CxFormat | CxRaw | CxDoubleQuote | CxString;
        _stringFlags["stringStartFRL"] = CxFormat | CxRaw | CxLong | CxString;
        _stringFlags["stringStartFRLD"] = CxFormat | CxRaw | CxLong | CxDoubleQuote | CxString;
    }

    private static readonly HashSet<string> BracketedNames =
    [
        "ParenthesizedExpression", "TupleExpression", "ComprehensionExpression", "importList", "ArgList", "ParamList",
        "ArrayExpression", "ArrayComprehensionExpression", "subscript",
        "SetExpression", "SetComprehensionExpression", "FormatString", "FormatReplacement", "nestedFormatReplacement",
        "DictionaryExpression", "DictionaryComprehensionExpression",
        "SequencePattern", "MappingPattern", "PatternArgList", "TypeParamList"
    ];

    private static int _termIndent, _termDedent, _termReplacementStart, _termString, _termFormatString;
    private static HashSet<int> _bracketOpens = new();
    private static Dictionary<int, int> _stringFlagById = new();

    private static void InitTermIds(Dictionary<string, int> terms)
    {
        _termIndent = terms["indent"];
        _termDedent = terms["dedent"];
        _termReplacementStart = terms["replacementStart"];
        _termString = terms["String"];
        _termFormatString = terms["FormatString"];
        _bracketOpens = new HashSet<int> { terms["ParenL"], terms["BracketL"], terms["BraceL"] };
        _stringFlagById = new Dictionary<int, int>();
        foreach (var (name, flag) in _stringFlags)
            _stringFlagById[terms[name]] = flag;
    }

    private static ContextTracker MakePythonTrackIndent()
    {
        return new ContextTracker(
            start: TopIndent,
            reduce: (context, term, stack, input) =>
            {
                var ctx = (PyContext)context!;
                var name = stack.Parser.TermNames?.GetValueOrDefault(term) ?? "";
                if ((ctx.Flags & CxBracketed) > 0 && BracketedNames.Contains(name) ||
                    (term == _termString || term == _termFormatString) && (ctx.Flags & CxString) > 0)
                    return ctx.Parent;
                return ctx;
            },
            shift: (context, term, stack, input) =>
            {
                var ctx = (PyContext)context!;
                if (term == _termIndent)
                    return new PyContext(ctx, CountIndent(input.Read(input.Pos, stack.Pos)), 0);
                if (term == _termDedent)
                    return ctx.Parent!;
                if (_bracketOpens.Contains(term) || term == _termReplacementStart)
                    return new PyContext(ctx, 0, CxBracketed);
                if (_stringFlagById.TryGetValue(term, out var sf))
                    return new PyContext(ctx, 0, sf | (ctx.Flags & CxBracketed));
                return ctx;
            },
            hash: (context) => ((PyContext)context!).Hash,
            @strict: false
        );
    }

    private static ITokenizer MakePythonExternalTokenizer(string name, Dictionary<string, int> terms)
    {
        if (name == "newlines")
        {
            var eof = terms["eof"];
            var newlineBracketed = terms["newlineBracketed"];
            var blankLineStart = terms["blankLineStart"];
            var newlineToken = terms["newline"];

            return new ExternalTokenizer((input, stack) =>
            {
                var ctx = (PyContext)stack.Context!;
                if (input.Next < 0)
                {
                    input.AcceptToken(eof);
                }
                else if ((ctx.Flags & CxBracketed) > 0)
                {
                    if (IsLineBreak(input.Next)) input.AcceptToken(newlineBracketed, endOffset: 1);
                }
                else if (((input.Peek(-1) < 0 || IsLineBreak(input.Peek(-1))) && stack.CanShift(blankLineStart)))
                {
                    var spaces = 0;
                    while (input.Next == Space || input.Next == Tab) { input.Advance(); spaces++; }
                    if (input.Next == Newline || input.Next == CarriageReturn || input.Next == Hash)
                        input.AcceptToken(blankLineStart, endOffset: -spaces);
                }
                else if (IsLineBreak(input.Next))
                {
                    input.AcceptToken(newlineToken, endOffset: 1);
                }
            }, contextual: true);
        }

        if (name == "indentation")
        {
            var indent = terms["indent"];
            var dedent = terms["dedent"];

            return new ExternalTokenizer((input, stack) =>
            {
                var ctx = (PyContext)stack.Context!;
                if (ctx.Flags != 0) return;
                var prev = input.Peek(-1);
                if (prev == Newline || prev == CarriageReturn)
                {
                    var depth = 0;
                    var chars = 0;
                    while (true)
                    {
                        if (input.Next == Space) depth++;
                        else if (input.Next == Tab) depth += 8 - (depth % 8);
                        else break;
                        input.Advance();
                        chars++;
                    }
                    if (depth != ctx.Indent &&
                        input.Next != Newline && input.Next != CarriageReturn && input.Next != Hash)
                    {
                        if (depth < ctx.Indent) input.AcceptToken(dedent, endOffset: -chars);
                        else input.AcceptToken(indent);
                    }
                }
            });
        }

        if (name == "legacyPrint")
        {
            var printKeyword = terms["printKeyword"];
            var printStr = "print";

            return new ExternalTokenizer((input, _) =>
            {
                for (var i = 0; i < 5; i++)
                {
                    if (input.Next != printStr[i]) return;
                    input.Advance();
                }
                if (char.IsLetterOrDigit((char)input.Next)) return;
                for (var off = 0; ; off++)
                {
                    var next = input.Peek(off);
                    if (next == Space || next == Tab) continue;
                    if (next != ParenOpen && next != Dot && next != Newline && next != CarriageReturn && next != Hash)
                        input.AcceptToken(printKeyword);
                    return;
                }
            });
        }

        if (name == "strings")
        {
            var stringContent = terms["stringContent"];
            var Escape = terms["Escape"];
            var replacementStart = terms["replacementStart"];
            var stringEnd = terms["stringEnd"];

            return new ExternalTokenizer((input, stack) =>
            {
                var ctx = (PyContext)stack.Context!;
                var flags = ctx.Flags;
                var quote = (flags & CxDoubleQuote) > 0 ? DoubleQuote : SingleQuote;
                var longStr = (flags & CxLong) > 0;
                var escapes = (flags & CxRaw) == 0;
                var format = (flags & CxFormat) > 0;

                var start = input.Pos;
                while (true)
                {
                    if (input.Next < 0)
                    {
                        break;
                    }
                    else if (format && input.Next == BraceOpen)
                    {
                        if (input.Peek(1) == BraceOpen)
                        {
                            input.Advance();
                            input.Advance();
                        }
                        else
                        {
                            if (input.Pos == start)
                            {
                                input.AcceptToken(replacementStart, endOffset: 1);
                                return;
                            }
                            break;
                        }
                    }
                    else if (escapes && input.Next == Backslash)
                    {
                        if (input.Pos == start)
                        {
                            input.Advance();
                            var escaped = input.Next;
                            if (escaped >= 0)
                            {
                                input.Advance();
                                SkipEscape(input, escaped);
                            }
                            input.AcceptToken(Escape);
                            return;
                        }
                        break;
                    }
                    else if (input.Next == Backslash && !escapes && input.Peek(1) > -1)
                    {
                        input.Advance();
                        input.Advance();
                    }
                    else if (input.Next == quote && (!longStr || (input.Peek(1) == quote && input.Peek(2) == quote)))
                    {
                        if (input.Pos == start)
                        {
                            input.AcceptToken(stringEnd, endOffset: longStr ? 3 : 1);
                            return;
                        }
                        break;
                    }
                    else if (input.Next == Newline)
                    {
                        if (longStr)
                        {
                            input.Advance();
                        }
                        else if (input.Pos == start)
                        {
                            input.AcceptToken(stringEnd);
                            return;
                        }
                        break;
                    }
                    else
                    {
                        input.Advance();
                    }
                }
                if (input.Pos > start) input.AcceptToken(stringContent);
            });
        }

        throw new ArgumentException($"Unknown Python external tokenizer: {name}");
    }

    private static void SkipEscape(InputStream input, int ch)
    {
        if (ch == LetterO)
        {
            for (var i = 0; i < 2 && input.Next >= 48 && input.Next <= 55; i++) input.Advance();
        }
        else if (ch == LetterX)
        {
            for (var i = 0; i < 2 && IsHex(input.Next); i++) input.Advance();
        }
        else if (ch == LetterU)
        {
            for (var i = 0; i < 4 && IsHex(input.Next); i++) input.Advance();
        }
        else if (ch == LetterBigU)
        {
            for (var i = 0; i < 8 && IsHex(input.Next); i++) input.Advance();
        }
        else if (ch == LetterN)
        {
            if (input.Next == BraceOpen)
            {
                input.Advance();
                while (input.Next >= 0 && input.Next != BraceClose && input.Next != SingleQuote &&
                       input.Next != DoubleQuote && input.Next != Newline)
                    input.Advance();
                if (input.Next == BraceClose) input.Advance();
            }
        }
    }

    private static readonly NodePropSource PythonHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["async \"*\" \"**\" FormatConversion FormatSpec"] = Tags.ModifierTag,
        ["for while if elif else try except finally return raise break continue with pass assert await yield match case"] = Tags.ControlKeyword,
        ["in not and or is del"] = Tags.OperatorKeyword,
        ["from def class global nonlocal lambda"] = Tags.DefinitionKeyword,
        ["import"] = Tags.ModuleKeyword,
        ["with as print"] = Tags.Keyword,
        ["Boolean"] = Tags.Bool,
        ["None"] = Tags.Null,
        ["VariableName"] = Tags.VariableName,
        ["CallExpression/VariableName"] = Tags.Function(Tags.VariableName),
        ["FunctionDefinition/VariableName"] = Tags.Function(Tags.Definition(Tags.VariableName)),
        ["ClassDefinition/VariableName"] = Tags.Definition(Tags.ClassName),
        ["PropertyName"] = Tags.PropertyName,
        ["CallExpression/MemberExpression/PropertyName"] = Tags.Function(Tags.PropertyName),
        ["Comment"] = Tags.LineComment,
        ["Number"] = Tags.Number,
        ["String"] = Tags.String,
        ["FormatString"] = Tags.Special(Tags.String),
        ["Escape"] = Tags.Escape,
        ["UpdateOp"] = Tags.UpdateOperator,
        ["ArithOp!"] = Tags.ArithmeticOperator,
        ["BitOp"] = Tags.BitwiseOperator,
        ["CompareOp"] = Tags.CompareOperator,
        ["AssignOp"] = Tags.DefinitionOperator,
        ["Ellipsis"] = Tags.Punctuation,
        ["At"] = Tags.Meta,
        ["( )"] = Tags.Paren,
        ["[ ]"] = Tags.SquareBracket,
        ["{ }"] = Tags.Brace,
        ["."] = Tags.DerefOperator,
        [", ;"] = Tags.Separator,
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var termTable = PythonParserData.TermTable!;
        BuildStringFlags(termTable);
        InitTermIds(termTable);
        var externals = new Dictionary<string, ITokenizer>
        {
            ["legacyPrint"] = MakePythonExternalTokenizer("legacyPrint", termTable),
            ["indentation"] = MakePythonExternalTokenizer("indentation", termTable),
            ["newlines"] = MakePythonExternalTokenizer("newlines", termTable),
            ["strings"] = MakePythonExternalTokenizer("strings", termTable),
        };
        var spec = PythonParserData.MakeSpec(
            externals: externals,
            propSources: [PythonHighlighting],
            context: MakePythonTrackIndent()
        );
        return new LRParser(spec);
    }
}
